File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.c
#include <stdio.h>
#include <string.h>
#define Min(x, y) ((x) < (y) ? (x) : (y))
#define Max(x, y) ((x) > (y) ? (x) : (y))

int main(void){
    int T, K;
    scanf("%d", &T);
    while(T--){
        int check[26][10001];
        char str[10001];
        scanf("%s", str);
        scanf("%d", &K);
        for(int i = 0; i < 26; i++) check[i][0] = 0;
        for(int i = 0; str[i]; i++){
            int idx = str[i] - 'a';
            check[idx][0]++;
            check[idx][check[idx][0]] = i;
        }

        int ans_min = 10001, ans_max = 0;
        for(int i = 0; i < 26; i++){
            if(check[i][0] < K) continue;
            int left = 1;
            int right = K;
            while(right <= check[i][0]){
                ans_max = Max(ans_max, check[i][right] - check[i][left] + 1);
                ans_min = Min(ans_min, check[i][right] - check[i][left] + 1);
                left++;
                right++;
            }
        }

        if(!ans_max) printf("-1\n");
        else printf("%d %d\n", ans_min, ans_max);
    }
}

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.py
import sys
input = sys.stdin.readline

T = int(input())
for _ in range(T):
    word = list(input().strip())
    K = int(input())
    ansMax = -1
    ansMin = 100000
    check = [[] for _ in range(26)]
    
    length = len(word)
    for i in range(length):
        check[ord(word[i]) - 97].append(i)
    
    for w in check:
        wlen = len(w)
        for i in range(wlen - K + 1):
            ansMax = max(ansMax, w[i + K - 1] - w[i] + 1)
            ansMin = min(ansMin, w[i + K - 1] - w[i] + 1)
    
    if ansMax == -1: print(-1)
    else: print(ansMin, ansMax)